### DISCUSSION TITLE
Overriding the injector when importing a module as a package

### DISCUSSION THREAD

**Discussion by cpurdy** [xtclang/xvm#37] [Ideas]
Splitting this out from the "syntax changes" thread into its own discussion thread:

>Overriding resource injection for imported modules (i.e. within the same type system). This is an already-decided project, but it just hasn't been on the critical path. Currently, all of the injections into a type system come from outside of the container, but when a module depends on (imports) another module, it may also desire to provide some of the injections into the depended-upon module itself. This would somehow have to be specified as part of the `package <name> import <module>` statement.

**Comment by cpurdy** [comment]
The basic idea is to allow the "package import module" statement to optionally specify a ResourceProvider for a specific set of type/name combinations:

    package Spring import springframework.spring.org
            supply (Clock wallClock, Console console, Console stderr) using SpringInjector;

The injector would be required to be a singleton `ResourceProvider` implementation, e.g.:

    static service SpringInjector
            implements ResourceProvider
        {
        Supplier getResource(Type type, String name)
            {
            return switch (type, name)
                {
                case (Clock, "wallClock"): // ...
                case (Console, "console"): // ...
                case (Console, "stderr"):  // ...
                default: assert;
                }
            }
        }

The resources enumerated as being overridden by the injector in the "package import module" statement would then be removed from the list of required injections of that module, and those injections within that module within the container being formed would be provided by the specified singleton `ResourceProvider` implementation.

See also:
* NestedContainer: public Set<InjectionKey> collectInjections()

Unresolved questions:
* What about evaluation order when different module import statements for the same module are reachable by different paths?

**Comment by ggleyzer** [comment]
Implemented
